### PR TITLE
Implement parametric reprovisioning and redeployment

### PR DESCRIPTION
**Pull request by guillett**
Depends on #7 and #8

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,9 +20,35 @@ BRANCH_NAME=master
 curl --location --remote-name https://github.com/sgmap/mes-aides-ops/archive/$BRANCH_NAME.tar.gz
 tar -xvf $BRANCH_NAME.tar.gz
 cd mes-aides-ops-$BRANCH_NAME
-./bootstrap.sh
+# ./bootstrap.sh mes-aides-ui_head mes-aides-ops_head
```
Comment (MattiSG) [inline README.md]: Why is this commented out? Either remove it or make the intention clearer.

File: README.md [hunk 2]
```diff
@@ -20,9 +20,35 @@ BRANCH_NAME=master
...
+./bootstrap.sh origin/master origin/$BRANCH_NAME
 ```
 
+`./bootstrap.sh origin/master origin/$BRANCH_NAME` initiates the set-up. By default, `origin/master` is used for both repositories (https://github.com/sgmap/mes-aides-ui and https://github.com/sgmap/mes-aides-ops). That can be overidden by passing *TREEISH* parameters. The fist one is `mes-aides-ui` head and the second one is `mes-aides-ops` one.
```
Comment (MattiSG) [inline README.md]: - fist → first
- Perhaps italic is enough for _treeish_? :wink:
- I don't understand what you mean by “head”. If I pass a treeish, then it's not necessarily `HEAD`.

File: README.md [hunk 3]
```diff
@@ -20,9 +20,35 @@ BRANCH_NAME=master
...
+
+That is why the suggested set of commands above overrides the default head of `mes-aides-ops` to rely on the selected branch.
+
+## Continuous provisioning and deployment
+
+### Provisioning
+
+A private key can `ssh` to the host and it will automatically trigger:
```
Comment (MattiSG) [inline README.md]: This is ambiguous: “a private key”? A specific one or any private key that can connect to the host?

File: Vagrantfile [hunk 4]
```diff
@@ -25,5 +25,6 @@ Vagrant.configure("2") do |config|
   config.vm.define "mes_aides_#{current_directory}"
   config.vm.network "private_network", ip: current_private_ip
 
-  config.vm.provision :shell, inline: '/vagrant/bootstrap.sh'
+  head_commit_hash = `cat .git/$(git symbolic-ref HEAD)`
```
Comment (MattiSG) [inline Vagrantfile]: This is really low-level. Couldn't you use `git rev-parse HEAD`?

File: bootstrap.sh [hunk 5]
```diff
@@ -32,6 +32,25 @@ do
...
+
+if [ $# -gt 0 ]
+then
+    ui_head=$1
+fi
+
+if [ $# -gt 1 ]
+then
+    ops_head=$2
+fi
```
Comment (MattiSG) [inline bootstrap.sh]: Replace all this with:

```sh
UI_TARGET_SHA=${1:-origin/master}
OPS_TARGET_SHA=${2:-origin/master}
```

🙂

File: modules/mesaides/manifests/automation_script.pp [hunk 6]
```diff
@@ -0,0 +1,13 @@
+define mesaides::automation_script (
```
Comment (MattiSG) [inline modules/mesaides/manifests/automation_script.pp]: Poor naming. We're in an automation machine. What does it do? `mesaides::update`?

**Comment by MattiSG** [comment]
After IRL discussion: it's probably best to simply use elements set inside the host at provisioning time to know which branches to use as sources. The added complexity of sending parameters in the suggested implementation doesn't seem to be compensated by the use cases.

File: modules/mesaides/files/update.sh [hunk 7]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+# Managed by puppet from automation_script.sh
```
Comment (MattiSG) [inline modules/mesaides/files/update.sh]: I'm not sure this script exists anywhere anymore.

File: modules/mesaides/files/update.sh [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+export PATH=/opt/puppetlabs/bin:$PATH
+
+echo "Your SSH connection ($1) triggered a shell script ($0)."
+case "$1" in
+    provision)
+        puppet apply $MES_AIDES_ROOT/manifests/ops.pp --verbose --modulepath=$MES_AIDES_ROOT/modules
+        puppet apply $MES_AIDES_ROOT/manifests/default.pp --verbose --modulepath=$MES_AIDES_ROOT/modules
+        ;;
+    deploy)
+        puppet apply $MES_AIDES_ROOT/manifests/default.pp --verbose --modulepath=$MES_AIDES_ROOT/modules
```
Comment (MattiSG) [inline modules/mesaides/files/update.sh]: > What's the added value of merging two dumb scripts into one with more intelligence and thus risks of mistakes?

**Comment by MattiSG** [comment]
Ok, I could connect. My mistake was not using `root` as a user, and afterwards having my SSH agent reusing a previously multiplexed connection with another key, thus preventing the deployment from being triggered. Congrats @guillett for the quick investigation!
`ssh -i ~/.ssh/deploy root@192.168.56.100` does trigger a deployment if I demux.

The previous comments still apply.

File: modules/mesaides/files/update.sh (hunk 8 above)
Comment (guillett) [inline modules/mesaides/files/update.sh]: Dumb*ish*, I like the fact that the logic for scripts triggered remotely is in one single place. It is still less than 25 lines. I keep as is for now and I rely on the review review to limit the risk of mistakes :-)

**Pull request closed without merging**